fv3gfs/util/halo_data_transformer.py

```diff
@@ -0,0 +1,905 @@
+        """Retrieve unpack buffer.
+
+        WARNING: Only available _after_ `compile` as been called.
+        """
+        if self._unpack_buffer is None:
+            raise RuntimeError("Recv buffer can't be retrieved before allocate()")
+        return self._unpack_buffer
+
+    def get_pack_buffer(self) -> Buffer:
+        """Retrieve pack buffer.
+
+        WARNING: Only available _after_ `compile` as been called.
+        """
+        if self._pack_buffer is None:
+            raise RuntimeError("Send buffer can't be retrieved before allocate()")
+        return self._pack_buffer
+
+    def _compile(self):
+        """Allocate contiguous memory buffers from description queued."""
+
+        # Compute required size
+        buffer_size = 0
+        dtype = None
+        for edge_x in self._infos_x:
+            buffer_size += edge_x._pack_buffer_size
+            dtype = edge_x.specification.dtype
+        if self._type is _HaloDataTransformerType.VECTOR:
+            for edge_y in self._infos_y:
+                buffer_size += edge_y._pack_buffer_size
+
+        # Retrieve two properly sized buffers
+        self._pack_buffer = Buffer.pop_from_cache(
+            self._np_module.zeros, (buffer_size), dtype
+        )
+        self._unpack_buffer = Buffer.pop_from_cache(
+            self._np_module.zeros, (buffer_size), dtype
+        )
+
+    def ready(self) -> bool:
+        """Check if the buffers are ready for communication."""
+        return self._pack_buffer is not None and self._unpack_buffer is not None
+
+    @abc.abstractclassmethod
+    def async_pack(
+        self,
+        quantities_x: List[Quantity],
+        quantities_y: Optional[List[Quantity]] = None,
+    ):
+        """Pack all given quantities into a single send Buffer.
+
+        Implementation should guarantee the send Buffer is ready for MPI communication.
+        The send Buffer is held by this class."""
+        raise NotImplementedError("Did you call HaloDataTransformer.get()?")
+
+    @abc.abstractclassmethod
+    def async_unpack(
+        self,
+        quantities_x: List[Quantity],
+        quantities_y: Optional[List[Quantity]] = None,
+    ):
+        """Unpack the buffer into destination quantities.
+
+        Implementation DOESN'T HAVE TO guarantee transfer is done."""
+        raise NotImplementedError("Did you call HaloDataTransformer.get()?")
+
+    @abc.abstractclassmethod
+    def finalize(self):
+        """Finalize operations after unpack.
+
+        Implementation should guarantee all transfers are done.
+        Implementation should return recv & send Buffers to the cache for reuse."""
+        raise NotImplementedError("Did you call HaloDataTransformer.get()?")
+
+    @abc.abstractclassmethod
+    def synchronize(self):
+        """Synchronize all operations.
+
+        Implementation guarantees all memory is now safe to access.
+        """
+        raise NotImplementedError("Did you call HaloDataTransformer.get()?")
```

Comment:
I believe these are instance methods, not class methods, so you'll want to use the `@abc.abstractmethod` decorator.

I suggested a re-write of these docstrings in a way that they transfer to the instances themselves. I haven't seen a python ABC's docstring written in a way that refers to "implementations" before. I also re-wrote them to read as I understand them, please double-check that the updated docstrings are accurate.

Because there are no routines which retrieve/initialize buffers other than init, the object is useless once `finalize` has been called. Given this is just clean-up, can `finalize` be renamed to `__del__` so we can have the clean-up occur automatically?

If `async_pack` guarantees the buffer is ready to read, is it really asynchronous?

Should we rely on the user to always call `synchronize` right before `get_unpack_buffer` or should we call `synchronize` inside `get_unpack_buffer` instead?

```suggestion
    @abc.abstractmethod
    def async_pack(
        self,
        quantities_x: List[Quantity],
        quantities_y: Optional[List[Quantity]] = None,
    ):
        """Pack all given quantities into a single send Buffer.

        Guarantees the buffer returned by `get_pack_buffer` is ready
        to be read and contains packed data.

        Args:
            quantities_x: scalar or vector x-component quantities to pack,
                if one is vector they must all be vector
            quantities_y: if quantities are vector, the y-component
                quantities.
        """
        pass

    @abc.abstractclassmethod
    def async_unpack(
        self,
        quantities_x: List[Quantity],
        quantities_y: Optional[List[Quantity]] = None,
    ):
        """Unpack the buffer into destination quantities.

        Does not guarantee the buffer returned by `get_unpack_buffer` has
        received data, doing so requires calling `synchronize`.

        Args:
            quantities_x: scalar or vector x-component quantities to pack,
                if one is vector they must all be vector
            quantities_y: if quantities are vector, the y-component
                quantities.
        """
        pass

    @abc.abstractclassmethod
    def finalize(self):
        """Return buffers to cache."""
        pass

    @abc.abstractclassmethod
    def synchronize(self):
        """Synchronize asynchronous operations.

        Guarantees get_pack_buffer and get_unpack_buffer contain any
        data resulting from asynchronous operations.
        """
        pass
```

Reply:
We did loose async behavior in the multiple rewrite of this. I'll introduce it back as this could have impact on later work & scaling.

As for the `__del__` I am always uneasy about using destructor/deleter path in garbage collected languages due to their inherent implicitness. But that's probably the python way to do things, so I'll move finalize in there.

Reply:
Alright I restored the asyncness where it was broken, moved a `synchronize` around the get_pack_buffer & get_unpack_buffer and updated the docstrings to reflect it all.

Got rid of `finalize_on_wait` entirely, relying on garbage collection to restore buffers